### ISSUE TITLE
cookbook for apps

### ISSUE THREAD

**Issue by KatherineCaley** [cogent3/cogent3#1636]
Apps need documentation!! This should be structured based on the type of apps, i.e., loaders, filtering steps, data sampling apps.

**Comment by GavinHuttley** [comment]
related to #1378

**Comment by GavinHuttley** [comment]
Following up on how to include code snippets, here is the [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html#examples) description of how they should be formatted.

**Comment by fredjaya** [comment]
Running `pytest` catches the code snippets in a simple example by adding:

- The code snippet in the docstring, with a pass and fail conditions
https://github.com/fredjaya/DemoProject/blob/ea776bc185b1cd252068619e2f0862a0d34b7c78/src/myproject/cli.py#L94-L101

- doctest options in the .toml
https://github.com/fredjaya/DemoProject/blob/ea776bc185b1cd252068619e2f0862a0d34b7c78/pyproject.toml?plain=1#L53-L55

Output is:
```
================ test session starts =================
platform linux -- Python 3.11.6, pytest-7.4.3, pluggy-1.3.0
rootdir: /home/frederickjaya/GitHub/DemoProject
configfile: pyproject.toml
testpaths: tests, src/myproject
plugins: anyio-4.1.0, cov-4.1.0, xdist-3.5.0
collected 3 items                                    

tests/test_myproject.py ..                     [ 66%]
src/myproject/cli.py F                         [100%]

====================== FAILURES ======================
_______ [doctest] myproject.cli._parse_csv_arg _______
095 
096     >>> _parse_csv_arg("test")
097     ['test']
098     >>> _parse_csv_arg("test")
Expected:
    ['fail']
Got:
    ['test']

/home/frederickjaya/mambaforge/envs/c311/lib/python3.11/site-packages/myproject/cli.py:98: DocTestFailure
============== short test summary info ===============
FAILED src/myproject/cli.py::myproject.cli._parse_csv_arg

```

**Comment by GavinHuttley** [comment]
Thanks @fredjaya ! I'm going to create another issue to trial this for one of the apps.

**Comment by fredjaya** [comment]
| module                | name                 | docstring | cookbook | assigned |
| --------------------- | -------------------- | --------- | -------- | -------- |
| cogent3.app.io        | pickle_it            |           |          |          |
| cogent3.app.io        | unpickle_it          |           |          |          |
| cogent3.app.io        | compress             |           |          |          |
| cogent3.app.io        | decompress           |           |          |          |
| cogent3.app.io        | to_primitive         |           |          |          |
| cogent3.app.io        | from_primitive       |           |          |          |
| cogent3.app.io        | to_json              |           |          |          |
| cogent3.app.io        | from_json            |           |          |          |
| cogent3.app.io        | load_aligned         |           |          |          |
| cogent3.app.io        | load_unaligned       |           |          |          |
| cogent3.app.io        | load_tabular         |           |          |          |
| cogent3.app.io        | load_json            |           |          |          |
| cogent3.app.io        | load_db              |           |          |          |
| cogent3.app.io        | write_json           |           |          |          |
| cogent3.app.io        | write_seqs           |           |          |          |
| cogent3.app.io        | write_tabular        |           |          |          |
| cogent3.app.io        | write_db             |           |          |          |
| cogent3.app.dist      | fast_slow_dist       |           |          |          |
| cogent3.app.dist      | jaccard_dist         |           |          |          |
| cogent3.app.dist      | approx_pdist         |           |          |          |
| cogent3.app.dist      | approx_jc69          |           |          |          |
| cogent3.app.dist      | gap_dist             |           |          |          |
| cogent3.app.tree      | scale_branches       |           |          |          |
| cogent3.app.tree      | uniformize_tree      |           |          |          |
| cogent3.app.tree      | quick_tree           |           |          |          |
| cogent3.app.align     | align_to_ref         |           |          |          |
| cogent3.app.align     | progressive_align    |           |          |          |
| cogent3.app.align     | smith_waterman       |           |          |          |
| cogent3.app.align     | ic_score             |           |          |          |
| cogent3.app.align     | cogent3_score        |           |          |          |
| cogent3.app.align     | sp_score             |           |          |          |
| cogent3.app.evo       | model                |           |          |          |
| cogent3.app.evo       | model_collection     |           |          |          |
| cogent3.app.evo       | hypothesis           |           |          |          |
| cogent3.app.evo       | bootstrap            |           |          |          |
| cogent3.app.evo       | ancestral_states     |           |          |          |
| cogent3.app.evo       | tabulate_stats       |           |          |          |
| cogent3.app.evo       | natsel_neutral       |           |          |          |
| cogent3.app.evo       | natsel_zhang         |           |          |          |
| cogent3.app.evo       | natsel_sitehet       |           |          |          |
| cogent3.app.evo       | natsel_timehet       |           |          |          |
| cogent3.app.translate | select_translatable  |           |          |          |
| cogent3.app.translate | translate_seqs       |           |          |          |
| cogent3.app.sample    | concat               |           |          |          |
| cogent3.app.sample    | omit_degenerates     |           |          |          |
| cogent3.app.sample    | omit_gap_pos         |           |          |          |
| cogent3.app.sample    | take_codon_positions |           |          |          |
| cogent3.app.sample    | take_named_seqs      |           |          |          |
| cogent3.app.sample    | take_n_seqs          |           |          |          |
| cogent3.app.sample    | min_length           |           |          |          |
| cogent3.app.sample    | fixed_length         |           |          |          |
| cogent3.app.sample    | omit_bad_seqs        |           |          |          |
| cogent3.app.sample    | omit_duplicated      |           |          |          |
| cogent3.app.sample    | trim_stop_codons     |           |          |          |

**Comment by KatherineCaley** [comment]
proposed new structure for apps, following the structure of the main documentation. 
` 
- apps 
  - overview 
  - cookbook
    - loaders
    - generic
    - ...
  - tutorials
    - ...
  - writing your own

**Comment by KatherineCaley** [comment]
@GavinHuttley @fredjaya 

*for loaders, it's hard to document in a docstring without including data with cogent3; for writers, it's similarly tricky. Should the examples not be strictly runnable but instead be indicative of how to run?

**Comment by fredjaya** [comment]
Thinking how it fits with the diataxis framework, will "cookbooks" be how-to guides? And is a reference/API section for apps required?

**Comment by GavinHuttley** [comment]
For the Q1 deadline:
1. focus on docstrings (since that goes out with the code).
2. cookbook (how-to's)
3. docs on "writing and publishing apps"

@KatherineCaley you are right about the docstring difficulty for loaders / writers. So those should probably use `.. code-block:: python` directives.

**Comment by KatherineCaley** [comment]
@fredjaya Yes, cookbooks are the how-to equivalent. I, too, am not sure about where API should go. The thesis for separating out the app documentation is that they are a distinct programming style. I've got two ideas (noting these are not necessary Q1 goals):
1. apps get there own API section 
2. apps are included in the general API (because they are cogent3 classes after all) and there is a link to this section from the app section

**Comment by GavinHuttley** [comment]
@GavinHuttley remove the apps that don't need to be shared, also in the .toml file

**Comment by YapengLang** [comment]
I'll help it out!

**Comment by fredjaya** [comment]
Note to check examples with `doctest-modules` once #1844 is addressed.

**Comment by KatherineCaley** [comment]
@GavinHuttley @fredjaya todo: Decide on the TOC location, whether it is hidden from the central document or not. Then, apply it systematically to the documents. 

Note that the current documents are not consistent!! 

see https://cogent3.org/doc/examples/index.html for example of central TOC 

see https://cogent3.org/doc/cookbook/index.html for example of LHS TOC only
